apkid/rules/elf/obfuscators.yara

```diff
@@ -130,7 +132,87 @@ rule ollvm_v6_0 : obfuscator
     $based_on      = "(based on Obfuscator-LLVM 6.0."
 
   condition:
-    all of them and not ollvm_v6_0_strenc
+    is_elf and
+    all of them and
+    not ollvm_v6_0_strenc
+}
+
+rule ollvm_v9_strenc : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM version 9.x (string encryption)"
+    sample      = "2314ec0053d829d424a82f702188fcb525cefce4feeef096f0855339b897a5d1"
+    author      = "Eduardo Novella"
+
+  strings:
+    $clang_version = /clang version \d\.\d\.\d /
+    $strenc        = /\.datadiv_decode[\d]{18,20}/
+    $ollvm         = "(based on Obfuscator-LLVM 9."
+
+  condition:
+    is_elf and all of them
+}
+
+rule ollvm_tll : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM TLL (string encryption)"
+    url         = "https://github.com/yazhiwang/ollvm-tll"
+    sample      = "1f010330e9ac90f00d11aa37fdca25c437ad6f4b1302f6d7aa48b91ef22cc107"
+    author      = "Eduardo Novella"
+
+  strings:
+    /**
+      .datadiv_decode7760209850571766755
+      Android clang version 5.0.300080  (based on LLVM 5.0.300080)
+      clang version 6.0.0 (tags/RELEASE_600/final) (https://github.com/yazhiwang/ollvm-tll.git a38559e4c13359073102793c0a734bb1add3d5ff)
+    */
+    $clang_version = /clang version \d\.\d\.\d \(tags\/RELEASE\_\d+\/final\)/
+    $strenc        = /\.datadiv_decode[\d]{18,20}/
+    $url           = "https://github.com/yazhiwang/ollvm-tll"
+
+  condition:
+    is_elf and all of them
+}
+
+rule ollvm_armariris : obfuscator
+{
+  meta:
+    description = "Armariris Obfuscator-LLVM (string encryption)"
+    url         = "https://github.com/GoSSIP-SJTU/Armariris"
+    sample      = "d22c2f53bab6fa2ab7bdb4f7acabb419e3ee3163bb758c4f7a013d07a8b09e12" // aka malware Joker
+    author      = "Eduardo Novella"
+
+  strings:
+    // clang version 3.9.1 (tags/RELEASE_391/final)
+    // clang version 5.0.1 (tags/RELEASE_501/final)
+    // .datadiv_decode14660921177804423408
+    $clang_version = /clang version \d\.\d\.\d \(tags\/RELEASE\_\d+\/final\)/
+    $strenc        = /\.datadiv_decode[\d]{18,20}/
+
+  condition:
+    is_elf and
+    all of them and
+    not ollvm_tll and
+    for any i in (0..elf.symtab_entries): (elf.symtab[i].name matches /\.datadiv_decode[\d]{18,20}/)
+}
+
+rule ollvm_strenc : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM version unknown (string encryption)"
+    sample      = "73f34f7dd5f5c2eff33fc48371c850a2a3ff0355a2bfa014467478ccb30309e3"
+    author      = "Eduardo Novella"
+
+  strings:
+    $strenc = /\.datadiv_decode[\d]{18,20}/
+
+  condition:
+    is_elf and $strenc and
+    not ollvm_tll and
+    not ollvm_armariris and
+    not ollvm_v6_0_strenc and
+    not ollvm_v9_strenc
 }
```

Comment:
nitpick, since the previous discussion was dismissed.

If we have a sample of this, which is currently the hash in the about. It seems that this isn't unknown.

IMHO - unknown should be what we expect to catch but don't have a sample of. E.g. - it uses the same encryption code, but fails to hit our other markers.

Since we have a sample of this, and know why it misses, it seems like a missed opportunity to classify this one correctly.

e.g. ;
if the strenc is found, and it's packed with UPX, just call it ollvm_strenc + UPX, unless you're not certain that is what it is?
This allows us to have another rule which is similar to what you have right now, which will actually only detect on unknown samples, right?

Reply:
Too many combinations so far. I'd like to merge and improve it later when users come back with better detection and more samples. Do you agree?